testplan/common/report/schemas.py

```diff
@@ -6,14 +6,58 @@
 
 from testplan.common.serialization import fields as custom_fields
 from testplan.common.serialization import schemas
+from testplan.common.utils import timing
 
-from .base import EventRecorder, Report, ReportGroup
+from .base import Report, BaseReportGroup, Status, RuntimeStatus
 
-__all__ = ["ReportLogSchema", "ReportSchema", "ReportGroupSchema"]
+__all__ = ["ReportLogSchema", "ReportSchema", "BaseReportGroupSchema"]
 
 # pylint: disable=unused-argument
 
 
+class IntervalSchema(Schema):
+    """Schema for ``timer.Interval``"""
+
+    start = custom_fields.UTCDateTime()
+    end = custom_fields.UTCDateTime(allow_none=True)
+
+    @post_load
+    def make_interval(self, data, **kwargs):
+        """Create an Interal object."""
+        return timing.Interval(**data)
+
+
+class TimerField(fields.Field):
+    """
+    Field for serializing ``timer.Timer`` objects, which is a ``dict``
+    of ``timer.Interval``.
+    """
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        return {
+            k: [IntervalSchema().dump(v) for v in l] for k, l in value.items()
+        }
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        return timing.Timer(
+            {
+                k: [IntervalSchema().load(v) for v in l]
+                for k, l in value.items()
+            }
+        )
+
+
+class ChildSchema(Schema):
+    """
+    Field for serializing ``timer.Timer`` objects, which is a ``dict``
```

Comment:
Yes, with `timer` I really meant `event`, I'm just re-using the name for backward compat, but we can change everything to event as well.

Currently the timer class is still a bit rough in terms of API, I will need to give it a bit more thoughts and design.

However, I kind of liked it that the timing data is embedded in the report hierarchy, so that we don't need to re-build the hierarchy somewhere else.

We can introduce logging level concept to timer recording later so that we can turn on/off with some config.

We can also have pub-sub for realtime update by defining a super-shallow schema of the report and only serialize e.g name & timer.

I see a potential of having this reverse coupling on the parent hierarchy, e.g for including driver information into multitest report.